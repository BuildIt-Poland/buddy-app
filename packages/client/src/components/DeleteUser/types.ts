import { UserRole } from '@buddy-app/schema';

export interface DeleteUserProps {
  userId: string;
  userRole: UserRole;
}
