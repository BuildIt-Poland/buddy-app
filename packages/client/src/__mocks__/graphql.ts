import { BUDDY_MENU_DETAILS, NEWBIE_MENU_DETAILS } from 'graphql/user-menu.graphql';
import { AVATAR_HEADER } from 'graphql/avatar-header.graphql';
import { NEWBIE_SELECT } from 'graphql/newbie-select.graphql';
import { BUDDY_SELECT } from 'graphql/buddy-select.graphql';
import { TASK_DETAILS } from 'graphql/task-details.graphql';
import { ADD_NEWBIE_TASK } from 'graphql/add-task.graphql';
import { ADD_NEWBIE } from 'graphql/add-user.graphql';
import { GraphQLError } from 'graphql';

import {
  MutationAddNewbieTaskArgs,
  AuthPayload,
  TaskStatus,
  Newbie,
  QueryNewbieArgs,
  QueryBuddyArgs,
  UserRole,
} from '@buddy-app/schema';
import {
  NEWBIE_CONTACT_DETAILS,
  BUDDY_CONTACT_DETAILS,
  TALENT_CONTACT_DETAILS,
} from 'graphql/contact-details.graphql';
import { TASK_LIST } from 'graphql/task-list.graphql';
import { getBasicUserDetailsMock } from './general';

export const loginSuccessMock = (data?: Partial<AuthPayload>) => ({
  data: {
    login: {
      role: 'BUDDY',
      token: 'dummy-token',
      userId: '1',
    },
  },
});

export const loginFailedMock = () => new GraphQLError('No such user found');

export const loginNoNetworkMock = () => ({
  networkError: true,
});

export const NewbieAvatarDetails = (variables?: QueryNewbieArgs) => ({
  request: {
    query: AVATAR_HEADER,
    variables,
  },
  result: {
    data: {
      newbie: {
        id: '1',
        name: 'Tom Hanks',
        position: 'front-end',
        photo: 'some-url',
      },
    },
  },
});

export const tasksResponse = (filter?: TaskStatus) => {
  const tasks = [
    {
      id: '1',
      title: 'Call you head manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Uncompleted,
      newbie: {} as Newbie,
    },
    {
      id: '2',
      title: 'Write you main manager',
      description: '',
      createdAt: '',
      implementationDate: '',
      status: TaskStatus.Completed,
      newbie: {} as Newbie,
    },
  ];

  switch (filter) {
    case TaskStatus.Completed:
      return [tasks[0]];
    case TaskStatus.Uncompleted:
      return [tasks[1]];
    default:
      return tasks;
  }
};

export const buddyMenuDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: {
          id: '1234',
          name: 'Tom Hanks',
          email: 'tom@wipro.com',
          allowPushedNotifications: true,
          photo: 'buddy-photo-url',
          newbies: [
            {
              id: 'id',
              name: 'Sandra Bullock',
              photo: 'newbie-photo-url',
            },
          ],
        },
      },
    },
  },
];

export const newbieMenuDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_MENU_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: {
          id: '1234',
          name: 'Sandra Bullock',
          email: 'doejoe@wipro.com',
          allowPushedNotifications: true,
          photo: 'newbie-photo-url',
          buddy: {
            id: 'id',
            name: 'Tom Hanks',
            photo: 'newbie-photo-url',
          },
        },
      },
    },
  },
];

export const taskListResponse = (
  variables: Partial<QueryNewbieArgs>,
  newbie: Partial<Newbie>
) => [
  {
    request: {
      query: TASK_LIST,
      variables,
    },
    result: {
      data: {
        newbie,
      },
    },
  },
];

export const newbieContactDetails = (variables?: Partial<QueryNewbieArgs>) => [
  {
    request: {
      query: NEWBIE_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        newbie: getBasicUserDetailsMock(UserRole.Newbie),
      },
    },
  },
];

export const buddyContactDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: BUDDY_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: getBasicUserDetailsMock(UserRole.Buddy),
      },
    },
  },
];

export const talentContactDetails = (variables?: Partial<QueryBuddyArgs>) => [
  {
    request: {
      query: TALENT_CONTACT_DETAILS,
      variables,
    },
    result: {
      data: {
        buddy: getBasicUserDetailsMock(UserRole.Talent),
      },
    },
  },
];

export const newbieSelectMock = [
  {
    request: {
      query: NEWBIE_SELECT,
      variables: {
        buddyId: 'ck17sl83c9gya0b17dcvttzm4',
      },
    },
    result: {
      data: {
        buddy: {
          id: 'ck17sl83c9gya0b17dcvttzm4',
          name: 'Dummy',
          role: 'BUDDY',
          photo: null,
          newbies: [
            {
              id: 'ck17svulh9k2k0b17j31ansfk',
              photo: null,
              name: 'Test 2',
              position: 'Front End',
              startDate: null,
              newbieTasks: [
                {
                  id: '1',
                  status: TaskStatus.Completed,
                },
              ],
              buddyTasks: [
                {
                  id: '1',
                  status: TaskStatus.Uncompleted,
                },
              ],
            },
            {
              id: 'ck17swp2m9kcv0b17we0ibrdn',
              photo: null,
              name: 'Test 1',
              position: 'Front End',
              startDate: null,
              newbieTasks: [
                {
                  id: '1',
                  status: TaskStatus.Completed,
                },
              ],
              buddyTasks: [
                {
                  id: '1',
                  status: TaskStatus.Uncompleted,
                },
              ],
            },
          ],
        },
      },
    },
  },
];

export const buddySelectMock = [
  {
    request: {
      query: BUDDY_SELECT,
      variables: {
        talentId: 'ck17sl83c9gya0b17dcvttzm4',
      },
    },
    result: {
      data: {
        talent: {
          id: 'ck17sl83c9gya0b17dcvttzm4',
          name: 'Dummy',
          role: 'BUDDY',
          photo: null,
          buddies: [],
        },
      },
    },
  },
];

export const taskDetailsMock = [
  {
    request: {
      query: TASK_DETAILS,
      variables: {
        taskId: '1',
      },
    },
    result: {
      data: {
        task: {
          id: '1',
          title: 'New task',
          description: 'New task',
          status: TaskStatus.Uncompleted,
          newbie: getBasicUserDetailsMock(),
        },
      },
    },
  },
];

export const addTaskSuccessMock = (
  variables?: Partial<MutationAddNewbieTaskArgs>,
  data?: Partial<AuthPayload>
) => ({
  request: {
    query: ADD_NEWBIE_TASK,
    variables: {
      newbieId: '1234',
      input: {
        title: 'Test task',
        description: '<h1>Hello world!</h1>',
      },
      ...variables,
    },
  },
  result: {
    data: {
      addNewbieTask: {
        id: 'ck3sw1h1mqkd90964u4675qo2',
        title: 'Test task',
        newbie: {
          id: '1',
          newbieTasks: [],
        },
        ...data,
      },
    },
  },
});

export const addTaskFailedMock = (
  variables?: Partial<MutationAddNewbieTaskArgs>
) => ({
  request: {
    query: ADD_NEWBIE_TASK,
    variables: {
      newbieId: '1234',
      input: {
        title: 'Test task',
        description: '<h1>Hello world!</h1>',
      },
      ...variables,
    },
  },
  result: {
    errors: [new GraphQLError('No such user')],
  },
});

export const addNewbieSuccessMock = () => ({
  request: {
    query: ADD_NEWBIE,
    variables: {
      buddyId: '1234',
      input: {
        email: 'test_newbie@email.com',
        name: 'test newbie',
      },
    },
  },
  result: {
    data: {
      addNewbie: {
        id: 'ck3sw1h1mqkd90964u4675qo2',
        email: 'test_newbie@email.com',
        name: 'test newbie',
        role: UserRole.Newbie,
      },
    },
  },
});

export const addNewbieFailedMock = () => ({
  request: {
    query: ADD_NEWBIE,
    variables: {
      buddyId: '1234',
      input: {
        email: 'test_newbie@email.com',
        name: 'test newbie',
      },
    },
  },
  result: {
    errors: [new GraphQLError('No such user')],
  },
});
