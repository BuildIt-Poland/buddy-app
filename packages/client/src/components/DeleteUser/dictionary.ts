interface Dictionary {
  DELETE_DIALOG: {
    TITLE: string;
    MESSAGE: string;
  };
  DELETE_SNACKBAR: {
    SUCCESS: string;
    ERROR: string;
  };
}

const DICTIONARY: Dictionary = {
  DELETE_DIALOG: {
    TITLE: 'Delete User',
    MESSAGE: `Are you sure you want to delete this user?
    All linked instances will be deleted as well.
    This action cannot be undone!`,
  },
  DELETE_SNACKBAR: {
    SUCCESS: 'User was successfully deleted',
    ERROR: 'An error ocurred deleting user',
  },
};

export default DICTIONARY;
