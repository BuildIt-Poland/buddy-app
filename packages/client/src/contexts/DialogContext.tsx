import React, { useState, useContext } from 'react';
import AlertDialog from 'atoms/AlertDialog';

interface State {
  isOpen: boolean;
  message: string;
  title: string | undefined;
  onConfirm?: () => void;
}

interface DialogContextData {
  showDialog: (message: string, title?: string, onConfirm?: () => void) => void;
  hideDialog: () => void;
}

const defaultState: State = {
  isOpen: false,
  title: '',
  message: '',
};

const DialogContext = React.createContext<DialogContextData | undefined>(undefined);

interface DialogProviderProps {
  children: React.ReactNode;
  value?: DialogContextData;
}

const DialogProvider = (props: DialogProviderProps): JSX.Element => {
  const [state, setDialog] = useState<State>(defaultState);

  const showDialog = (message: string, title?: string, onConfirm?: () => void) =>
    setDialog({ isOpen: true, message, title, onConfirm });

  const hideDialog = () => setDialog(defaultState);

  const onConfirmHandler = () => {
    state.onConfirm && state.onConfirm();
    hideDialog();
  };

  return (
    <DialogContext.Provider
      value={{
        showDialog,
        hideDialog,
      }}>
      <AlertDialog
        isOpen={state.isOpen}
        message={state.message}
        onConfirm={state.onConfirm ? onConfirmHandler : undefined}
        title={state.title}
        onClose={hideDialog}
      />
      {props.children}
    </DialogContext.Provider>
  );
};

const useDialog = () => {
  const context = useContext(DialogContext);
  if (context === undefined) {
    throw new Error(`useDialog must be used within a DialogProvider`);
  }
  return context;
};

export { DialogProvider, useDialog };
