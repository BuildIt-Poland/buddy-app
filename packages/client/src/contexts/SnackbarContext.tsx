import React, { useState, useContext } from 'react';
import SnackBar from 'atoms/SnackBar';

interface State {
  isOpen: boolean;
  message: string;
}

interface SnackbarProviderProps {
  children: React.ReactNode;
  value?: State;
}
interface SnackbarContextData extends State {
  showSnackbar: (message: string) => void;
  hideSnackBar: () => void;
}

const defaultState: State = {
  isOpen: false,
  message: '',
};

const SnackbarContext = React.createContext<SnackbarContextData | undefined>(
  undefined
);

const SnackbarProvider = ({
  children,
  value,
}: SnackbarProviderProps): JSX.Element => {
  const [{ isOpen, message }, setSnackBarState] = useState<State>(
    value || defaultState
  );

  const showSnackbar = (message: string) =>
    setSnackBarState(state => ({ ...state, isOpen: true, message }));

  const hideSnackBar = () =>
    setSnackBarState(state => ({ ...state, isOpen: false }));

  return (
    <SnackbarContext.Provider
      value={{
        isOpen,
        message,
        showSnackbar,
        hideSnackBar,
      }}>
      <SnackBar isOpen={isOpen} message={message} onClose={hideSnackBar} />
      {children}
    </SnackbarContext.Provider>
  );
};

const useSnackBar = () => {
  const context = useContext(SnackbarContext);

  if (context === undefined) {
    throw new Error(`useSnackBar must be used within a SnackbarContext`);
  }
  return context;
};

export { SnackbarProvider, useSnackBar };
