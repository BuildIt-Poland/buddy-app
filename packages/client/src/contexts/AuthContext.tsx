import React, { useReducer, useEffect, useContext } from 'react';
import { Mutation, AuthPayload } from '@buddy-app/schema';
import { useQuery } from '@apollo/react-hooks';
import { ExecutionResult } from '@apollo/react-common';
import { auth } from 'utils';
import { LOGIN_MUTATION } from 'graphql/login.graphql';
import { GET_LOCAL_AUTH } from 'graphql/get-auth.graphql';
import { AuthCache, setCacheToken } from 'utils/apollo-client/cache';
import { apolloClient } from 'utils';

enum ActionTypes {
  AUTH_INIT = 'auth/init',
  AUTH_ERROR = 'auth/error',
  AUTH_SUCCESS = 'auth/success',
  AUTH_LOGOUT = 'auth/logout',
}

interface Action {
  type: ActionTypes;
  payload?: AuthPayload;
}

export interface AuthState {
  loading: boolean;
  error?: any;
  data: AuthPayload;
  isAuthenticated: boolean;
}

const defaultState: AuthState = {
  loading: false,
  error: null,
  data: {} as AuthPayload,
  isAuthenticated: false,
};

interface AuthContext extends AuthState {
  login: (email: string, password: string) => Promise<ExecutionResult<Mutation>>;
  logout: () => void;
}

interface AuthProviderProps {
  children: React.ReactNode;
  value?: AuthContext;
}

const defaultContext: AuthContext = {
  ...defaultState,
  login: () => Promise.reject(),
  logout: () => null,
};

const AuthContext = React.createContext(defaultContext);

const authReducer = (state: AuthState, action: Action): AuthState => {
  switch (action.type) {
    case ActionTypes.AUTH_INIT:
      return {
        ...state,
        loading: true,
        error: null,
      };
    case ActionTypes.AUTH_SUCCESS:
      return {
        ...state,
        loading: false,
        data: { ...action.payload } as AuthPayload,
        isAuthenticated: true,
      };
    case ActionTypes.AUTH_ERROR:
      return {
        ...state,
        loading: false,
        error: { ...action.payload },
      };
    case ActionTypes.AUTH_LOGOUT:
      return {
        ...state,
        data: {} as AuthPayload,
        isAuthenticated: false,
      };
    default:
      throw new Error('Not valid action');
  }
};

const AuthProvider = ({ children, ...props }: AuthProviderProps) => {
  const [state, dispatch] = useReducer(authReducer, defaultState);
  const { data } = useQuery<AuthCache>(GET_LOCAL_AUTH);

  useEffect(() => {
    const user = auth.getUser();
    if (user) {
      dispatch({
        type: ActionTypes.AUTH_SUCCESS,
        payload: user,
      });
    }
  }, []);

  useEffect(() => {
    if (data && data.auth && data.auth.tokenHasExpired) {
      logout();
    }
  }, [data]);

  const login = async (email: string, password: string) => {
    dispatch({ type: ActionTypes.AUTH_INIT });

    try {
      const { data } = await apolloClient.mutate<Mutation>({
        mutation: LOGIN_MUTATION,
        variables: {
          email,
          password,
        },
      });

      if (data) {
        auth.setUser(data.login);
        dispatch({
          type: ActionTypes.AUTH_SUCCESS,
          payload: data.login,
        });
      }

      return Promise.resolve({ data });
    } catch (error) {
      dispatch({ type: ActionTypes.AUTH_ERROR, payload: error });
      return Promise.reject(error);
    }
  };

  const logout = () => {
    auth.removeUser();
    apolloClient.writeData({
      data: {
        ...setCacheToken(false),
      },
    });
    dispatch({ type: ActionTypes.AUTH_LOGOUT });
  };

  return (
    <AuthContext.Provider value={{ ...state, login, logout }} {...props}>
      {children}
    </AuthContext.Provider>
  );
};

const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error(`useAuth must be used within a AuthProvider`);
  }
  return context;
};

export { AuthProvider, useAuth };
