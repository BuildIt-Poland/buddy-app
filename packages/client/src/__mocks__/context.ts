import { AuthState } from 'contexts/AuthContext';
import { UserRole } from '.';

export const mockedTalentContext = (context?: Partial<AuthState>) => ({
  isAuthenticated: true,
  loading: false,
  data: {
    role: UserRole.Talent,
    token: 'token',
    userId: '1234',
  },
  ...context,
});

export const mockedBuddyContext = (context?: Partial<AuthState>) => ({
  isAuthenticated: true,
  loading: false,
  data: {
    role: UserRole.Buddy,
    token: 'token',
    userId: '1234',
  },
  ...context,
});

export const mockedNewbieContext = (context?: Partial<AuthState>) => ({
  isAuthenticated: true,
  loading: false,
  data: {
    role: UserRole.Newbie,
    token: 'token',
    userId: '1234',
  },
  ...context,
});
