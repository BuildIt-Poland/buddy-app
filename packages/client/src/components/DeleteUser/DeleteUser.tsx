import React from 'react';
import { useParams } from 'react-router-dom';
import { QueryBuddyArgs } from '@buddy-app/schema';
import RemoveCircleOutline from '@material-ui/icons/RemoveCircleOutline';
import IconButton from '@material-ui/core/IconButton';
import { useAuth } from 'contexts/AuthContext';
import { useDialog } from 'contexts/DialogContext';
import { useSnackBar } from 'contexts/SnackbarContext';
import useUserDelete from 'hooks/useUserDelete';
import { DeleteUserProps } from './types';
import DICTIONARY from './dictionary';
/* istanbul ignore file */
const DeleteUser: React.FC<DeleteUserProps> = ({ userId, userRole }) => {
  const { buddyId } = useParams<QueryBuddyArgs>();
  const { data } = useAuth();
  const { showDialog } = useDialog();
  const { showSnackbar } = useSnackBar();

  const [deleteUser] = useUserDelete(buddyId || data.userId, data.role, userRole, {
    onCompleted: () => showSnackbar(DICTIONARY.DELETE_SNACKBAR.SUCCESS),
    onError: () => showSnackbar(DICTIONARY.DELETE_SNACKBAR.ERROR),
  });

  const onDeleteHandler = (event: React.MouseEvent) => {
    event.preventDefault();
    showDialog(
      DICTIONARY.DELETE_DIALOG.MESSAGE,
      DICTIONARY.DELETE_DIALOG.TITLE,
      () => deleteUser(userId)
    );
  };

  return (
    <IconButton aria-label='remove-user' onClick={onDeleteHandler}>
      <RemoveCircleOutline />
    </IconButton>
  );
};

export default DeleteUser;
