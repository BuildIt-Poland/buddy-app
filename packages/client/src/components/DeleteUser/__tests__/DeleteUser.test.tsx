import React from 'react';
import { create } from 'react-test-renderer';
import { MemoryRouter, Route } from 'react-router';
import { MockedProvider } from '@apollo/react-testing';
import { AuthProvider } from 'contexts/AuthContext';
import { SnackbarProvider } from 'contexts/SnackbarContext';
import { DialogProvider } from 'contexts/DialogContext';
import { UserRole, mockedBuddyContext } from '__mocks__';
import DeleteUser from '../DeleteUser';

jest.mock('@material-ui/icons/RemoveCircleOutline', () => 'RemoveCircleOutline');
jest.mock('@material-ui/core/IconButton', () => 'IconButton');

describe('Component - DeleteUser', () => {
  test('renders correctly', () => {
    const path = '/buddy/newbies';

    const component = create(
      <MockedProvider addTypename={false} resolvers={{}}>
        <MemoryRouter initialEntries={[path]}>
          <Route path={'/buddy/newbies/:newbieId/tasks'}>
            <DialogProvider>
              <SnackbarProvider>
                <AuthProvider value={mockedBuddyContext()}>
                  <DeleteUser userId='1' userRole={UserRole.Newbie} />
                </AuthProvider>
              </SnackbarProvider>
            </DialogProvider>
          </Route>
        </MemoryRouter>
      </MockedProvider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
