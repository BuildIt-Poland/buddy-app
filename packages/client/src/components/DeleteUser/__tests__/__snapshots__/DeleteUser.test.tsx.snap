// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - UserGrid renders correctly Should render correctly UserGrid 1`] = `
<Grid
  className="makeStyles-grid-4"
  component="ul"
  container={true}
  justify="center"
  wrap="wrap"
>
  <Grid
    className="makeStyles-gridItem-3"
    component="li"
    item={true}
  >
    <a
      className="makeStyles-link-5"
      href="/talent/buddies/ck17svulh9k2k0b17j31ansfk/newbies"
      onClick={[Function]}
    >
      <Card
        className="makeStyles-card-1"
      >
        <CardContent
          className="makeStyles-cardInfo-2"
        >
          <Avatar
            imgSrc={null}
            name="Test 2"
            position="Front End"
          />
        </CardContent>
      </Card>
    </a>
  </Grid>
  <Grid
    className="makeStyles-gridItem-3"
    component="li"
    item={true}
  >
    <a
      className="makeStyles-link-5"
      href="/talent/buddies/ck17swp2m9kcv0b17we0ibrdn/newbies"
      onClick={[Function]}
    >
      <Card
        className="makeStyles-card-1"
      >
        <CardContent
          className="makeStyles-cardInfo-2"
        >
          <Avatar
            imgSrc={null}
            name="Test 1"
            position="Front End"
          />
        </CardContent>
      </Card>
    </a>
  </Grid>
</Grid>
`;
